scivision/catalog/update.py

```diff
@@ -0,0 +1,166 @@
+import requests
+import base64
+
+
+def _get_catalog(url: str):
+    """Load the Scivision catalog from GitHub.
+
+    Parameters
+    ----------
+    url : str
+        A github raw url pointing to the scivision catalog file.
+
+    Returns
+    -------
+    catalog_dict : dict
+        Dictionary version of scivision catalog.
+    """
+    resp = requests.get(url)
+    catalog_dict = json.loads(resp.text)
+    return catalog_dict
+
+
+def _update_catalog(entry: str, catalog_dict: dict, catalog: str = 'github', type: str = 'data') -> dict:
+    """Add a new entry to a catalog.
+
+    Parameters
+    ----------
+    entry : str
+        A path to the json file specifying the entry to be added to the catalog.
+    catalog_dict : dict
+        A dict containing the scivision catalog.
+    catalog : str
+        A path to the json file containing the scivision model catalog.
+        When 'github', updates the scivision catalog on GitHub.
+    type : str
+        A string that instructs which catalog to update, 'dataset' by default, can be changed to 'model'
+    """
+
+    # Get a dict for the new catalog entry
+    with open(entry) as file:
+        entry_dict = json.load(file)
+
+    # Check that an entry of this name doesn't already exist in the catalog
+    for key in entry_dict.keys():
+        if key in catalog_dict:
+            raise KeyError('Entry named: "' + key + '" already in catalog')
+
+    # Add the new entry to the catalog dict
+    updated_catalog_dict = catalog_dict | entry_dict  # Note: Python 3.9+ only
+
+    # Save the modified catalog to github or to file
+    if catalog == 'github':
+        catalog_json_string = json.dumps(updated_catalog_dict, sort_keys=True, indent=4)
+        _launch_pull_request(catalog_json_string, type=type)
+    else:
+        with open(catalog, 'w') as old_catalog:
+            json.dump(updated_catalog_dict, old_catalog, sort_keys=True, indent=4)
+
+
+def _launch_pull_request(catalog_json: str, type: str = 'data') -> None:
+    """Sends a pull request to the scivision-catalog repo, adding to the scivision catalog.
+    Parameters
+    ----------
+    catalog_json : str
+        A string containing an updated scivision catalog in json format.
+    type : str
+        A string that instructs which catalog to update, 'dataset' by default, can be changed to 'model'
+    """
+    catalog_repo = 'alan-turing-institute/scivision-catalog'
+    if type == 'data':
+        catalog_name = 'datasources'
+    elif type == 'model':
+        catalog_name = 'models'
+    catalog_file = catalog_name + ".json"
+
+    # Configure PyGitHub
+    print('Provide your GitHub access token (or hit enter to use default):')
+    token = input()
+    # if the user does not enter a token, use a default
+    # this is the token of a GH account called scivisioncontributor
+    # this account can fork and PR the catalog on the user's behalf
+    if token == "":
+        token = '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'
```

Comment:
Hi @edwardchalstrey1 - Just wanted to flag this - having a token in the source like this is generally a bad idea from a security perspective. Anyone who has it can act as the service user to perform certain actions, that could go well beyond what you had intended below. It seems like this token has been revoked (just tried using it, but would suggest making sure that this is the case!)

Before carrying on with this approach, please could we pause/rethink?

Reply:
To add to the above - regardless of the security implications, it also means all of these commits come from the service user - we should make sure people can get credit for their contributions, to their own GitHub account!

Reply:
yes, just testing this idea out to see just how automated we can make it.

I think the roadblock of this approach is that if you have to create a github token in order to use the function, this is scarcely any easier than creating a manual PR, which makes the add functions a but pointless. Hence I experimented with a workaround, but agree this approach doesn't work nicely! The token I created for `scivisioncontributor` keeps expiring despite it being set to not expire, suggesting that GH doesn't like tokens being used in this way!